## Подготовка

### Подключение библиотек

In [27]:
import numpy as np
import pandas as pd
from pandas_profiling import ProfileReport
import matplotlib.pyplot as plt
from ext.utilities import *
from ext.models import *

### Загрузка набора данных

In [28]:
dataset_path = './dataset/googleplaystore.csv'
df = pd.read_csv(dataset_path)

### Исследование базовых данных

In [29]:
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [31]:
df.describe()

Rating
count  9367.000000
mean      4.193338
std       0.537431
min       1.000000
25%       4.000000
50%       4.300000
75%       4.500000
max      19.000000

### Описание датасета

    Количество строк:        10841
    Количество столбцов:     13

     НАЗВАНИЕ   |      ОПИСАНИЕ        |   ТИП ПЕРЕМЕННОЙ  |   ТИП ДАННЫХ  |   ЧИСЛО ПУСТЫХ ЗНАЧЕНИЙ
    ------------+----------------------+-------------------+---------------+------------------------
       App      | название приложения  |     метрическая   |    string     |        0
                |                      |                   |               |
     Category   | Категория, к которой |    номинальная    |    string     |        0
                | относится приложение |                   |               |
                |                      |                   |               |
      Rating    | Общий рейтинг поль-  |                   |               |
                | зователя приложения  |    метрическая    |     float     |      1474
                |  (как при очистке)   |                   |               |
                |                      |                   |               |
                |  Количество отзывов  |                   |               |
      Reviews   |     пользователей    |    метрическая    |      int      |        0
                |      о приложении    |                   |               |
                |                      |                   |               |
       Size     |  Размер приложения   |   метрическая     |     string    |        0
                |                      |                   |               |
                | Количество пользо-   |                   |               |
     Installs   | вательских загрузок/ |    метрическая    |     string    |        0
                | установок приложения |                   |               |
                |                      |                   |               |
       Type     |     Платный или      |    номинальная    |     string    |        1
                |      бесплатный      |                   |               |
                |                      |                   |               |
      Price     |        Цена          |     метрическая   |     float     |        0
                |                      |                   |               |
                |  Возрастная группа,  |                   |               |
     Content    | на которую ориентиро |    номинальная    |     string    |        1
     Rating     | вано приложение Дети |                   |               |
                | /Зрелые 21+/Взрослый |                   |               |
                |                      |                   |               |
     Genres     |        Жанр          |    номинальная    |    string     |        0
                |                      |                   |               |
                | Дата последнего об-  |                   |               |
    LastUpdated | новления приложения  |    метрическая    |    datetime   |        0
                |    в Play Store      |                   |               |
                |                      |                   |               |
                | Текущая версия при-  |                   |               |
    Current Ver | ложения, достпупная  |    метрическая    |     string    |        8
                |    в Play Store      |                   |               |
                |                      |                   |               |
                |      Минимальная     |                   |               |
    Android Ver |   требуемая версия   |    метрическая    |     string    |        3
                |        Android       |                   |               |
                |                      |                   |               |

In [32]:
#profile = ProfileReport(df, title="Отчет о датасете")
#profile

### Изменение данных

In [33]:
# меняем дату вида January 15, 2018 к формату python datetime
df['Last Updated'] = list(map(parse_datetime, df['Last Updated']))

In [34]:
# меняем число установок
df['Installs'] = list(map(parse_float, df['Installs']))

In [35]:
# меняем число цен
df['Price'] = list(map(parse_float, df['Price']))

In [36]:
# столбец заработанных денег
df['Income'] = df['Price'] * df['Installs']
df['Income']

0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
        ... 
10836    0.0
10837    0.0
10838    0.0
10839    0.0
10840    0.0
Name: Income, Length: 10841, dtype: float64

In [37]:
# столбец рейтинга, переведенного из метрики в категорию модели Rating
df['Rating'] = list(map(get_rating, df['Rating']))

In [38]:
#Сформируйте срез по таблице и транспонируйте его.
df[220:229].T

220  \
App             104 Looking for a job - looking for a job, loo...   
Category                                                 BUSINESS   
Rating                                      Rating.UNSATISFACTORY   
Reviews                                                     74359   
Size                                                          25M   
Installs                                                1000000.0   
Type                                                         Free   
Price                                                         0.0   
Content Rating                                           Everyone   
Genres                                                   Business   
Last Updated                                  2018-07-27 00:00:00   
Current Ver                                                1.12.5   
Android Ver                                            4.0 and up   
Income                                                        0.0   

                                  221                           222  \
App                     Myanmar 2D/3D  Quick PDF Scanner + OCR FREE   
Category                     BUSINESS                      BUSINESS   
Rating          Rating.UNSATISFACTORY         Rating.UNSATISFACTORY   
Reviews                           822                         80805   
Size                             3.1M            Varies with device   
Installs                     100000.0                     5000000.0   
Type                             Free                          Free   
Price                             0.0                           0.0   
Content Rating               Everyone                      Everyone   
Genres                       Business                      Business   
Last Updated      2018-06-08 00:00:00           2018-02-26 00:00:00   
Current Ver                       1.5            Varies with device   
Android Ver              4.0.3 and up                  4.0.3 and up   
Income                            0.0                           0.0   

                                  223                             224  \
App                              sABN  ATI Cargoes and Transportation   
Category                     BUSINESS                        BUSINESS   
Rating          Rating.UNSATISFACTORY           Rating.UNSATISFACTORY   
Reviews                          2287                            4162   
Size                             1.5M                            7.5M   
Installs                    1000000.0                        100000.0   
Type                             Free                            Free   
Price                             0.0                             0.0   
Content Rating               Everyone                        Everyone   
Genres                       Business                        Business   
Last Updated                      NaT             2018-07-24 00:00:00   
Current Ver                       4.9                           0.9.4   
Android Ver                2.2 and up                    4.0.3 and up   
Income                            0.0                             0.0   

                                  225                    226  \
App                     Secure Folder             UPS Mobile   
Category                     BUSINESS               BUSINESS   
Rating          Rating.UNSATISFACTORY  Rating.UNSATISFACTORY   
Reviews                         14760                  23243   
Size                             8.6M     Varies with device   
Installs                   50000000.0              5000000.0   
Type                             Free                   Free   
Price                             0.0                    0.0   
Content Rating               Everyone               Everyone   
Genres                       Business               Business   
Last Updated      2018-01-31 00:00:00    2018-06-25 00:00:00   
Current Ver                  1.1.07.6     Varies with device   
Android Ver                7.0 a

In [39]:
#Отсортируйте значения по нескольким полям (многоуровневая сортировка) и выведите 20 первых строк таблицы, содержащих столбцы с переменными, по которым проводилась сортировка.
df.sort_values(['Installs','Last Updated'], ascending=[False, False])
df[['Installs','Last Updated']].head(20)

Installs Last Updated
0      10000.0   2018-01-07
1     500000.0   2018-01-15
2    5000000.0   2018-08-01
3   50000000.0   2018-06-08
4     100000.0   2018-06-20
5      50000.0          NaT
6      50000.0   2018-04-26
7    1000000.0   2018-06-14
8    1000000.0   2017-09-20
9      10000.0   2018-07-03
10   1000000.0   2017-10-27
11   1000000.0   2018-07-31
12  10000000.0   2018-04-02
13    100000.0   2018-06-26
14    100000.0   2018-08-03
15      5000.0   2018-06-06
16    500000.0   2018-07-31
17     10000.0   2017-11-07
18   5000000.0   2018-08-03
19  10000000.0   2018-07-30